File: app/src/main/java/com/mamba/picme/features/idphoto/IDPhotoScreen.kt
package com.mamba.picme.features.idphoto

import androidx.compose.foundation.Canvas
import androidx.compose.foundation.background
import androidx.compose.foundation.gestures.detectDragGestures
import androidx.compose.foundation.gestures.detectTransformGestures
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.rounded.Check
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Scaffold
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.collectAsState
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableFloatStateOf
import androidx.compose.runtime.mutableStateListOf
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.produceState
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.StrokeCap
import androidx.compose.ui.graphics.asImageBitmap
import androidx.compose.ui.graphics.drawscope.Stroke
import androidx.compose.ui.input.pointer.pointerInput
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.unit.IntOffset
import androidx.compose.ui.unit.IntSize
import androidx.compose.ui.unit.dp
import com.mamba.picme.R
import com.mamba.picme.domain.matting.IDPhotoComposer
import com.mamba.picme.domain.matting.IDPhotoSpecs
import com.mamba.picme.domain.matting.StrokeMode
import com.mamba.picme.features.common.topbar.AppTopBar
import com.mamba.picme.features.common.topbar.AppTopBarAction
import com.mamba.picme.features.idphoto.components.ColorSwatchRow
import com.mamba.picme.features.idphoto.components.EdgePanel
import com.mamba.picme.features.idphoto.components.IdPhotoTabRow
import com.mamba.picme.features.idphoto.components.RepairPanel
import com.mamba.picme.features.idphoto.components.RepairPanelCallbacks
import com.mamba.picme.features.idphoto.components.RepairPanelState
import com.mamba.picme.features.idphoto.components.SizeChipRow

@Suppress("LongMethod") // 待重构：抽 IDPhoto 控制面板子组件
@Composable
fun IDPhotoScreen(
    sourceUri: String,
    viewModel: IDPhotoViewModel,
    onNavigateBack: () -> Unit,
    onSaved: () -> Unit
) {
    val context = LocalContext.current
    val state by viewModel.state.collectAsState()

    LaunchedEffect(sourceUri) {
        viewModel.load(context, sourceUri)
        viewModel.onSaveComplete = { onSaved() }
    }

    Scaffold(
        topBar = {
            AppTopBar(
                title = stringResource(R.string.id_photo_title),
                onBack = onNavigateBack,
                actions = {
                    val ready = state as? IDPhotoViewModel.State.Ready
                    AppTopBarAction(
                        icon = Icons.Rounded.Check,
                        contentDescription = stringResource(R.string.done),
                        onClick = { viewModel.save(context) },
                        enabled = ready != null && !ready.isSaving
                    )
                }
            )
        }
    ) { padding ->
        Box(
            modifier = Modifier
                .fillMaxSize()
                .padding(padding)
                .background(Color(0xFF101010)),
            contentAlignment = Alignment.Center
        ) {
            when (val s = state) {
                is IDPhotoViewModel.State.Loading -> CircularProgressIndicator(color = MaterialTheme.colorScheme.primary)
                is IDPhotoViewModel.State.Error -> Text(s.message, color = Color.White, modifier = Modifier.padding(16.dp))
                is IDPhotoViewModel.State.Ready -> {
                    // 底图在加载/换底色/参数或描边变化时重建；手势只改 cropRect 绘制窗口，保证跟手
                    val base by produceState<android.graphics.Bitmap?>(
                        initialValue = null,
                        s.selectedColorIndex, s.edgeParams, s.strokeVersion
                    ) {
                        value = viewModel.previewBase()
                    }
                    // 修补 tab 的本地交互态（画框坐标系，仅用于实时覆盖层与笔刷光标）
                    var brushMode by remember { mutableStateOf(StrokeMode.ERASE) }
                    var brushSize by remember { mutableFloatStateOf(32f) }
                    var softEdge by remember { mutableStateOf(false) }
                    val overlayPoints = remember { mutableStateListOf<Offset>() }
                    var cursor by remember { mutableStateOf<Offset?>(null) }

                    // 重建期间（base 被重置为 null）保留上一帧底图，避免 spinner 闪断与覆盖层消失
                    var lastBase by remember { mutableStateOf<android.graphics.Bitmap?>(null) }
                    LaunchedEffect(base) {
                        if (base != null) {
                            lastBase = base
                            overlayPoints.clear() // 新底图就绪后无缝交接覆盖层（无覆盖层时 no-op）
                        }
                    }
                    val shownBase = base ?: lastBase

                    Column(
                        modifier = Modifier.fillMaxSize().padding(16.dp),
                        verticalArrangement = Arrangement.spacedBy(16.dp),
                        horizontalAlignment = Alignment.CenterHorizontally
                    ) {
                        Box(
                            modifier = Modifier
                                .weight(1f)
                                .fillMaxWidth(),
                            contentAlignment = Alignment.Center
                        ) {
                            val bmp = shownBase
                            val cropRect = viewModel.currentCropRect()
                            if (bmp != null && cropRect != null) {
                                val sizeSpec = IDPhotoSpecs.SIZES[s.selectedSizeIndex]
                                val frameW = 220.dp
                                val frameH = frameW * sizeSpec.heightPx / sizeSpec.widthPx
                                val repairing = s.activeTab == IdPhotoTab.REPAIR
                                Box(
                                    modifier = Modifier
                                        .size(frameW, frameH)
                                        .background(Color.White, RoundedCornerShape(4.dp))
                                        .clip(RoundedCornerShape(4.dp))
                                        .then(
                                            if (repairing) {
                                                // 不变量：REPAIR tab 不会改变 cropRect（涂抹期间禁缩放），故描边中途不会因 key 变化重启手势块
                                                Modifier.pointerInput(cropRect) {
                                                    try {
                                                        detectDragGestures(
                                                            onDragStart = { start ->
                                                                overlayPoints.clear()
                                                                cursor = start
                                                                val radiusSrc = IDPhotoComposer.frameRadiusToSource(
                                                                    brushSize / 2f, size.width.toFloat(), cropRect
                                                                )
                                                                viewModel.beginStroke(
                                                                    brushMode, radiusSrc,
                                                                    if (softEdge) 0.5f else 0f
                                                                )
                                                                overlayPoints.add(start)
                                                                viewModel.appendStrokePoint(
                                                                    IDPhotoComposer.frameToSource(
                                                                        px = start.x, py = start.y,
                                                                        frameW = size.width.toFloat(),
                                                                        frameH = size.height.toFloat(),
                                                                        crop = cropRect
                                                                    )
                                                                )
                                                            },
                                                            onDrag = { change, _ ->
                                                                change.consume()
                                                                cursor = change.position
                                                                overlayPoints.add(change.position)
                                                                viewModel.appendStrokePoint(
                                                                    IDPhotoComposer.frameToSource(
                                                                        px = change.position.x,
                                                                        py = change.position.y,
                                                                        frameW = size.width.toFloat(),
                                                                        frameH = size.height.toFloat(),
                                                                        crop = cropRect
                                                                    )
                                                                )
                                                            },
                                                            onDragEnd = {
                                                                cursor = null
                                                                viewModel.endStroke()
                                                            },
                                                            onDragCancel = {
                                                                cursor = null
                                                                overlayPoints.clear()
                                                                viewModel.endStroke()
                                                            }
                                                        )
                                                    } finally {
                                                        // 协程被取消（切 tab / pointerInput 重启）时 detectDragGestures 不走 onDragCancel，兜底收尾
                                                        if (cursor != null) {
                                                            cursor = null
                                                            overlayPoints.clear()
                                                            viewModel.endStroke()
                                                        }
                                                    }
                                                }
                                            } else {
                                                Modifier.pointerInput(Unit) {
                                                    detectTransformGestures { _, pan, gestureZoom, _ ->
                                                        viewModel.transformBy(
                                                            dxFraction = pan.x / size.width,
                                                            dyFraction = pan.y / size.height,
                                                            zoomChange = gestureZoom
                                                        )
                                                    }
                                                }
                                            }
                                        )
                                ) {
                                    val imageBmp = remember(bmp) { bmp.asImageBitmap() }
                                    Canvas(modifier = Modifier.fillMaxSize()) {
                                        drawImage(
                                            image = imageBmp,
                                            srcOffset = IntOffset(cropRect.left, cropRect.top),
                                            srcSize = IntSize(
                                                cropRect.right - cropRect.left,
                                                cropRect.bottom - cropRect.top
                                            ),
                                            dstSize = IntSize(size.width.toInt(), size.height.toInt())
                                        )
                                        // 进行中的描边覆盖层：恢复=白、擦除=黑，40% 透明
                                        if (overlayPoints.isNotEmpty()) {
                                            val overlayColor =
                                                (if (brushMode == StrokeMode.RESTORE) Color.White else Color.Black)
                                                    .copy(alpha = 0.4f)
                                            for (i in 1 until overlayPoints.size) {
                                                drawLine(
                                                    color = overlayColor,
                                                    start = overlayPoints[i - 1],
                                                    end = overlayPoints[i],
                                                    strokeWidth = brushSize,
                                                    cap = StrokeCap.Round
                                                )
                                            }
                                        }
                                        // 笔刷圈光标
                                        cursor?.let { pos ->
                                            drawCircle(
                                                color = Color.White.copy(alpha = 0.8f),
                                                radius = brushSize / 2f,
                                                center = pos,
                                                style = Stroke(width = 2f)
                                            )
                                        }
                                    }
                                }
                            } else {
                                CircularProgressIndicator(color = MaterialTheme.colorScheme.primary)
                            }
                        }
                        Text(
                            text = stringResource(
                                if (s.activeTab == IdPhotoTab.REPAIR) R.string.id_photo_repair_hint
                                else R.string.id_photo_drag_hint
                            ),
                            style = MaterialTheme.typography.bodySmall,
                            color = Color.White.copy(alpha = 0.6f)
                        )
                        IdPhotoTabRow(selected = s.activeTab, onSelect = viewModel::selectTab)
                        when (s.activeTab) {
                            IdPhotoTab.BG_COLOR ->
                                ColorSwatchRow(IDPhotoSpecs.COLORS, s.selectedColorIndex, viewModel::selectColor)
                            IdPhotoTab.SIZE ->
                                SizeChipRow(IDPhotoSpecs.SIZES, s.selectedSizeIndex, viewModel::selectSize)
                            IdPhotoTab.EDGE ->
                                EdgePanel(
                                    params = s.edgeParams,
                                    onParamsChange = viewModel::setEdgeParams,
                                    onReset = viewModel::resetEdgeParams
                                )
                            IdPhotoTab.REPAIR ->
                                RepairPanel(
                                    state = RepairPanelState(
                                        mode = brushMode,
                                        brushSizePx = brushSize,
                                        softEdge = softEdge,
                                        canUndo = s.canUndoStroke,
                                        canRedo = s.canRedoStroke,
                                        hasStrokes = s.canUndoStroke || s.canRedoStroke
                                    ),
                                    callbacks = RepairPanelCallbacks(
                                        onModeChange = { brushMode = it },
                                        onBrushSizeChange = { brushSize = it },
                                        onSoftEdgeChange = { softEdge = it },
                                        onUndo = viewModel::undoStroke,
                                        onRedo = viewModel::redoStroke,
                                        onClear = viewModel::clearStrokes
                                    )
                                )
                        }
                    }
                }
            }
        }
    }
}

File: iosApp/PoLang/Features/IdPhoto/IDPhotoScreen.swift
import CoreGraphics
import SwiftUI
import UIKit

/// 证照屏（对标 specs/screens/idphoto.yaml §1/§2/§6.4/§7/§8；定稿截图 tmp/ui-reference/idphoto-02..05）。
/// fullScreenCover 全屏强制深色；Loading/Error/Ready 三态（§7.1）。
/// Ready = [弹性预览框 + 构图提示行 + Tab 行 + 面板]（§2，padding/spacing 均 spacing.lg=16）。
/// 🔴 预览渲染契约（§2.1 render_contract）：底图按 currentCropRect 裁切后 Canvas 精确拉伸填满整框，
/// 禁止 Image+scale/translation 组合（历史白边 bug 235281d5b）；重算期间保留上一帧防闪白。
struct IDPhotoScreen: View {
    let localIdentifier: String

    @StateObject private var vm: IDPhotoViewModel
    @Environment(\.dismiss) private var dismiss

    // 合成底图（nil=尚未就绪；重算期间保留上一帧显示——§2.1 rebuild_retention）
    @State private var lastBase: CGImage?

    // 平移/缩放增量结算（§6.4：保存上次值传增量，连续拖动连续生效）
    @State private var lastDragTranslation: CGSize = .zero
    @State private var lastMagnification: CGFloat = 1

    // 修补画笔工具参数（§6；住在 Screen——跨 tab 切换保留，绑定进 RepairPanel）
    @State private var brushMode: StrokeMode = .erase
    @State private var brushSize: Float = RepairConstants.defaultBrushSize
    @State private var softEdge = false
    // 进行中描边 overlay（框内像素坐标）+ 光标（仅拖动中显示）
    @State private var overlayPoints: [CGPoint] = []
    @State private var cursor: CGPoint?

    // 保存失败 toast（§7.5；成功走 vm.onSaved → dismiss）
    @State private var saveError: String?

    private var ready: IdPhotoState.Ready? {
        if case .ready(let r) = vm.state { return r } else { return nil }
    }

    init(localIdentifier: String) {
        self.localIdentifier = localIdentifier
        _vm = StateObject(wrappedValue: IDPhotoViewModel(localIdentifier: localIdentifier))
    }

    var body: some View {
        ZStack {
            IdPhotoTokens.screenBackground.ignoresSafeArea()
            VStack(spacing: 0) {
                topBar
                stateContent
            }
            if let saveError {
                toast(saveError)
            }
        }
        .preferredColorScheme(.dark)   // §8 强制深色（cover 不继承根 scheme，必须显式）
        .task {
            vm.onSaved = { dismiss() }
            vm.onSaveFailed = { message in
                saveError = message
                DispatchQueue.main.asyncAfter(deadline: .now() + 1.6) {
                    if saveError == message { saveError = nil }
                }
            }
        }
    }

    // MARK: - 顶栏（§1：48pt 通用顶栏；back=退出不保存，check=保存）

    private var topBar: some View {
        AppTopBar(title: L("id_photo_title"),
                  showsBackButton: true,
                  onBack: { dismiss() }) {
            AppTopBarAction(systemName: "mat_o_check",
                            accessibilityID: "idphoto_save",
                            isEnabled: ready.map { readyState in !readyState.isSaving } ?? false) {
                Task { await vm.save() }
            }
            .accessibilityLabel(L("Done"))   // spec §1 actions desc=done
        }
    }

    // MARK: - 内容区（§1：按状态单选）

    @ViewBuilder
    private var stateContent: some View {
        switch vm.state {
        case .loading:
            ProgressView()
                .tint(AppColorScheme.dark.primary)
                .frame(maxWidth: .infinity, maxHeight: .infinity)
        case .error(let message):
            Text(message)
                .foregroundStyle(.white)
                .padding(Spacing.lg)
                .frame(maxWidth: .infinity, maxHeight: .infinity)
        case .ready(let ready):
            readyContent(ready)
        }
    }

    // MARK: - Ready 布局（§2：预览区 weight(1f) + 提示行 + Tab 行 + 面板）

    private func readyContent(_ ready: IdPhotoState.Ready) -> some View {
        VStack(spacing: Spacing.lg) {
            previewFrame(ready)
                .frame(maxWidth: .infinity, maxHeight: .infinity)   // 垂直弹性区，内容居中
            Text(L(ready.activeTab == .repair ? "id_photo_repair_hint" : "id_photo_drag_hint"))
                .font(AppTypography.bodySmall.font)
                .foregroundStyle(.white.opacity(AppAlpha.secondary))
            IdPhotoTabRow(selected: ready.activeTab, onSelect: { tab in vm.selectTab(tab) })
            panel(for: ready)
        }
        .padding(Spacing.lg)
        .frame(maxWidth: .infinity, maxHeight: .infinity)
        // 底图刷新：task(id:) 必然重跑（onChange 在 TabView 页面有不触发前科）；nil 时保留旧帧
        .task(id: baseKey(ready)) {
            guard let image = await vm.previewBase() else { return }
            lastBase = image
            overlayPoints = []   // 新底图就绪才替换并清空修补 overlay 点（§2.1）
            cursor = nil
        }
    }

    /// 底图内容指纹（colorIndex + edgeParams + strokeVersion，§7 preview 缓存 key 同源）
    private func baseKey(_ ready: IdPhotoState.Ready) -> String {
        "\(ready.selectedColorIndex)-\(ready.edgeParams)-\(ready.strokeVersion)"
    }

    // MARK: - 预览框（§2.1：白边回归契约 + §6.4 手势切换）

    private func previewFrame(_ ready: IdPhotoState.Ready) -> some View {
        let spec = sizeSpec(at: ready.selectedSizeIndex)
        let frameHeight = IdPhotoTokens.frameWidth * CGFloat(spec.pixelH) / CGFloat(spec.pixelW)
        let crop = vm.currentCropRect()
        return previewCanvas(crop: crop)
            .frame(width: IdPhotoTokens.frameWidth, height: frameHeight)
            .background(IdPhotoTokens.frameBackground)
            .clipShape(RoundedRectangle(cornerRadius: IdPhotoTokens.frameRadius))
            .contentShape(Rectangle())
            .gesture(frameGesture(activeTab: ready.activeTab, frameHeight: frameHeight))
            .simultaneousGesture(magnifyGesture(activeTab: ready.activeTab))
    }

    /// Canvas 渲染：合成底图按 cropRect 裁切后**精确拉伸填满整框**（零留边零白条）+ 修补 overlay
    private func previewCanvas(crop: CropRect?) -> some View {
        Canvas { context, size in
            if let base = lastBase, let crop {
                let cropRect = CGRect(x: crop.left, y: crop.top, width: crop.width, height: crop.height)
                var tile = base
                if cropRect.width > 0, cropRect.height > 0, let cropped = base.cropping(to: cropRect) {
                    tile = cropped
                }
                context.draw(Image(uiImage: UIImage(cgImage: tile)), in: CGRect(origin: .zero, size: size))
            }
            drawStrokeOverlay(context: &context)
        }
    }

    /// 修补进行中 overlay（§6 in_progress_overlay）：restore=白/erase=黑 alpha 0.4 圆头折线，
    /// 线宽=brushSize（框内像素）；光标=白 0.8 圆圈描边（宽 2，spec §6 定值），仅拖动中显示
    private func drawStrokeOverlay(context: inout GraphicsContext) {
        guard !overlayPoints.isEmpty else { return }
        let paintColor = (brushMode == .restore ? Color.white : Color.black).opacity(AppAlpha.placeholder)
        let brushWidth = CGFloat(brushSize)
        if overlayPoints.count == 1 {
            let p = overlayPoints[0]
            let r = brushWidth / 2
            context.fill(Path(ellipseIn: CGRect(x: p.x - r, y: p.y - r, width: r * 2, height: r * 2)),
                         with: .color(paintColor))
        } else {
            var path = Path()
            path.move(to: overlayPoints[0])
            for point in overlayPoints.dropFirst() {
                path.addLine(to: point)
            }
            context.stroke(path, with: .color(paintColor),
                           style: StrokeStyle(lineWidth: brushWidth, lineCap: .round, lineJoin: .round))
        }
        if let c = cursor {
            let r = brushWidth / 2
            context.stroke(Path(ellipseIn: CGRect(x: c.x - r, y: c.y - r, width: r * 2, height: r * 2)),
                           with: .color(Color.white.opacity(AppAlpha.emphasis)),
                           lineWidth: 2)
        }
    }

    // MARK: - 面板区（§2.4：按 tab 单选显示）

    @ViewBuilder
    private func panel(for ready: IdPhotoState.Ready) -> some View {
        switch ready.activeTab {
        case .bgColor:
            ColorSwatchRow(selectedIndex: ready.selectedColorIndex,
                           onSelect: { index in vm.selectColor(index) })
        case .size:
            SizeChipRow(selectedIndex: ready.selectedSizeIndex,
                        onSelect: { index in vm.selectSize(index) })
        case .edge:
            EdgePanel(edgeParams: ready.edgeParams,
                      onCommit: { params in vm.setEdgeParams(params) },
                      onReset: { vm.resetEdgeParams() })
        case .repair:
            RepairPanel(brushMode: $brushMode,
                        brushSize: $brushSize,
                        softEdge: $softEdge,
                        canUndo: ready.canUndoStroke,
                        canRedo: ready.canRedoStroke,
                        hasStrokes: ready.hasStrokes,
                        onUndo: { vm.undoStroke() },
                        onRedo: { vm.redoStroke() },
                        onClear: { vm.clearStrokes() })
        }
    }

    // MARK: - 手势（§6.4 / §6 painting_contract：REPAIR=单指画笔，其余=拖拽平移+双指缩放）

    private func frameGesture(activeTab: IdPhotoTab, frameHeight: CGFloat) -> some Gesture {
        DragGesture(minimumDistance: 0)
            .onChanged { value in
                if activeTab == .repair {
                    paintChanged(value, frameHeight: frameHeight)
                } else {
                    panChanged(value, frameHeight: frameHeight)
                }
            }
            .onEnded { value in
                if activeTab == .repair {
                    paintEnded()
                } else {
                    lastDragTranslation = .zero
                }
            }
    }

    /// 平移：增量结算（本次 translation − 上次 translation），保证连续拖动连续生效；
    /// 符号语义（右拖=取景窗左移）由 VM 内部处理，UI 原样传。
    /// ⚠️ 视图失效依赖：VM 的 offsetX/zoom 非 @Published——本函数对 lastDragTranslation 的
    /// @State 写入（及 magnifyGesture 对 lastMagnification 的写入）承担了驱动重渲染的职责，
    /// 二者不可「清理」掉，否则拖拽静默失灵（审查 Y4 登记）。
    private func panChanged(_ value: DragGesture.Value, frameHeight: CGFloat) {
        let dx = value.translation.width - lastDragTranslation.width
        let dy = value.translation.height - lastDragTranslation.height
        lastDragTranslation = value.translation
        vm.transformBy(dxFraction: Float(dx / IdPhotoTokens.frameWidth),
                       dyFraction: Float(dy / frameHeight),
                       zoomChange: 1)
    }

    /// 缩放：增量因子（value / 上次值，>1=放大）；双指期间 Drag 同时触发——增量结算下自然兼容
    private func magnifyGesture(activeTab: IdPhotoTab) -> some Gesture {
        MagnificationGesture()
            .onChanged { value in
                guard activeTab != .repair, lastMagnification > 0 else { return }
                vm.transformBy(dxFraction: 0, dyFraction: 0, zoomChange: Float(value / lastMagnification))
                lastMagnification = value
            }
            .onEnded { _ in lastMagnification = 1 }
    }

    /// 画笔：拖动开始 beginStroke（半径仅按宽度轴换算——屏幕上圆形笔刷），
    /// 移动逐点 append（源图坐标，domain 纯函数换算）；crop 为 nil 时不落笔
    private func paintChanged(_ value: DragGesture.Value, frameHeight: CGFloat) {
        let p = value.location
        if overlayPoints.isEmpty, let crop = vm.currentCropRect() {
            let radiusSource = IDPhotoComposer.frameRadiusToSource(
                radiusPx: brushSize / 2,
                frameW: Float(IdPhotoTokens.frameWidth),
                crop: crop)
            vm.beginStroke(mode: brushMode,
                           radiusSourcePx: radiusSource,
                           softness: softEdge ? RepairConstants.softnessOn : RepairConstants.softnessOff)
        }
        overlayPoints.append(p)
        cursor = p
        if let crop = vm.currentCropRect() {
            let source = IDPhotoComposer.frameToSource(
                px: Float(p.x), py: Float(p.y),
                frameW: Float(IdPhotoTokens.frameWidth), frameH: Float(frameHeight),
                crop: crop)
            vm.appendStrokePoint(xSourcePx: source.x, ySourcePx: source.y)
        }
    }

    /// 画笔结束：无条件 endStroke（手势取消也走这里——§6 cancel_safety）
    private func paintEnded() {
        overlayPoints = []
        cursor = nil
        vm.endStroke()
    }

    // MARK: - 工具

    private func sizeSpec(at index: Int) -> IDPhotoSizeSpec {
        let all = IDPhotoSizeSpec.allCases
        return all.indices.contains(index) ? all[index] : all[0]
    }

    /// 保存失败 toast（PhotoEditorScreen 先例样式）
    private func toast(_ message: String) -> some View {
        Text(message)
            .font(AppTypography.bodyMedium.font)
            .foregroundStyle(.white)
            .padding(.horizontal, Spacing.lg)
            .padding(.vertical, Spacing.sm)
            .background(Color.black.opacity(AppAlpha.emphasis))
            .clipShape(Capsule())
            .transition(.opacity)
    }
}
